File: task_3_1_single_thread.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

static double get_seconds(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}

long long count_inside_circle(long long samples, unsigned int seed) {
    long long inside = 0;
    unsigned int state = seed;
    for (long long i = 0; i < samples; ++i) {
        double x = (double)rand_r(&state) / RAND_MAX * 2.0 - 1.0;
        double y = (double)rand_r(&state) / RAND_MAX * 2.0 - 1.0;
        if (x*x + y*y <= 1.0) {
            inside++;
        }
    }
    return inside;
}

int main(void) {
    const long long num_samples = 10000000LL;
    const int num_runs = 5;
    double total_time = 0.0;
    double min_time = 1e9;
    double max_time = 0.0;
    double avg_pi = 0.0;

    printf("Running single-threaded approach with %lld samples...\n", num_samples);
    printf("Running %d times to compute average...\n\n", num_runs);

    for (int run = 0; run < num_runs; ++run) {
        double start = get_seconds();
        long long inside = count_inside_circle(num_samples, 42 + run);
        double elapsed = get_seconds() - start;
        double estimated_pi = (double)inside / (double)num_samples * 4.0;
        avg_pi += estimated_pi;
        total_time += elapsed;
        if (elapsed < min_time) min_time = elapsed;
        if (elapsed > max_time) max_time = elapsed;
        printf("Run %d: time = %.6f s, pi = %.8f\n", run + 1, elapsed, estimated_pi);
    }

    avg_pi /= num_runs;
    double avg_time = total_time / num_runs;
    double accuracy = (1.0 - fabs(avg_pi - M_PI) / M_PI) * 100.0;

    printf("\n=============================================================\n");
    printf("BASELINE (SINGLE-THREADED) RESULTS\n");
    printf("=============================================================\n");
    printf("Number of samples: %lld\n", num_samples);
    printf("Number of runs: %d\n", num_runs);
    printf("Average time: %.6f s\n", avg_time);
    printf("Min time: %.6f s\n", min_time);
    printf("Max time: %.6f s\n", max_time);
    printf("Estimated Pi: %.8f\n", avg_pi);
    printf("True Pi: %.8f\n", M_PI);
    printf("Accuracy: %.4f%%\n", accuracy);
    printf("=============================================================\n");
    return 0;
}

File: task_3_1_single_thread.py
"""
Task 3.1 - Approach 1: Single Thread (Baseline)
This version establishes a reference for performance comparison.

Requirements:
- Generate a large number of points (10^7 or higher)
- Count how many fall inside the circle
- Measure total execution time
- Use a fixed random seed to ensure reproducibility
- Avoid printing inside loops (I/O distorts timing)
- Run the program multiple times and compute the average execution time
"""

import random
import time
import math


def monte_carlo_pi_single_thread(num_samples):
    """
    Single-threaded Monte Carlo estimation of Pi.
    
    Args:
        num_samples: Number of random points to generate
    
    Returns:
        tuple: (estimated_pi, execution_time)
    """
    random.seed(42)  # Fixed seed for reproducibility
    inside_circle = 0
    
    # Generate random points and count those inside the circle
    for _ in range(num_samples):
        x = random.uniform(-1, 1)
        y = random.uniform(-1, 1)
        
        if x**2 + y**2 <= 1:
            inside_circle += 1
    
    # Estimate Pi: (inside_circle / num_samples) * 4
    estimated_pi = (inside_circle / num_samples) * 4
    
    return estimated_pi


def measure_performance(num_samples, num_runs=5):
    """
    Measure the performance of single-threaded approach over multiple runs.
    
    Args:
        num_samples: Number of random points to generate
        num_runs: Number of times to run the experiment
    
    Returns:
        dict: Statistics including average time, min, max, and estimated Pi
    """
    times = []
    pi_estimates = []
    
    print(f"Running single-threaded approach with {num_samples:,} samples...")
    print(f"Running {num_runs} times to compute average...")
    
    for run in range(num_runs):
        start_time = time.time()
        estimated_pi = monte_carlo_pi_single_thread(num_samples)
        elapsed_time = time.time() - start_time
        
        times.append(elapsed_time)
        pi_estimates.append(estimated_pi)
    
    avg_time = sum(times) / len(times)
    min_time = min(times)
    max_time = max(times)
    avg_pi = sum(pi_estimates) / len(pi_estimates)
    accuracy = (1 - abs(avg_pi - math.pi) / math.pi) * 100
    
    return {
        'num_samples': num_samples,
        'num_runs': num_runs,
        'average_time': avg_time,
        'min_time': min_time,
        'max_time': max_time,
        'all_times': times,
        'estimated_pi': avg_pi,
        'accuracy': accuracy,
        'true_pi': math.pi
    }


if __name__ == "__main__":
    # Generate a large number of points for baseline measurement
    num_samples = 10**7  # 10 million points
    
    results = measure_performance(num_samples, num_runs=5)
    
    # Print results
    print("\n" + "="*70)
    print("BASELINE (SINGLE-THREADED) RESULTS")
    print("="*70)
    print(f"Number of samples:    {results['num_samples']:,}")
    print(f"Number of runs:       {results['num_runs']}")
    print(f"Average execution time: {results['average_time']:.6f} seconds")
    print(f"Min execution time:     {results['min_time']:.6f} seconds")
    print(f"Max execution time:     {results['max_time']:.6f} seconds")
    print(f"Std deviation:          {(max(results['all_times']) - min(results['all_times'])) / 2:.6f} seconds")
    print(f"\nEstimated Pi:         {results['estimated_pi']:.8f}")
    print(f"True Pi:              {results['true_pi']:.8f}")
    print(f"Accuracy:             {results['accuracy']:.4f}%")
    print("="*70)
